files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py

```diff
@@ -0,0 +1,299 @@
+
+    logging.info(f'JSON file is {args.pin_config_filename}')
+
+    if args.request_pin:
+        tag, typ = args.update_pin
+        request_pin(tag, typ)
+
+    if args.update_pin:
+        tag, pin = args.update_pin
+        update_pin(args.pin_config_filename, tag, pin)
+
+    if args.get_pin:
+        value = get_pin(args.pin_config_filename, args.get_pin)
+        if value is not None:
+            print(f'The value associated with {args.get_pin} is {value}')
+        else:
+            print(f'No value associated with {args.get_pin} was found')
+
+    if args.reset_config:
+        reset(args.pin_config_filename)
+
+    if args.clear_unused:
+        clear_unused(args.pin_config_filename)
+
+
+def update_pin(pin_config_filename, tag, pin):
+    """Updates this tag to use the specified pin
+
+    This method has no effect if the tag is not already managed or if the new pin is a different
+    type.
+
+    Args:
+        tag (str): tag to update
+        pin (str): new physical pin id
+    """
+    logging.info(f'Updating pin {tag}')
+    with open(pin_config_filename, 'r+') as f:
+        pin_config = json.load(f)
+        if tag not in pin_config:
+            logging.info(f'Pin {tag} not found - not updating')
+            return
+        for pin in pin_table:
+            if pin[0] == pin:
+                if pin[2] != pin_config[tag]['type']:
+                    logging.info(f'Pin {tag} is a different type - not updating')
+                else:
+                    pin_config[tag]['pin'] = pin
+                    f.seek(0) # should reset file position to the beginning.
+                    json.dump(pin_config, f, indent=4)
+                    f.truncate() # remove remaining part
+                    logging.info(f'Pin {tag} updated to use {pin}')
+                    return
+
+    logging.info(f'{pin} is not a valid pin')
+
+
+def get_pin(pin_config_filename, tag):
+    """Return the value associated with the key in the json file.
+
+    Args:
+        tag (str): tag to search for
+
+    Returns:
+        str: value associated with tag
+    """
+    logging.info(f'Getting pin {tag}')
+    value = None
+    with open(pin_config_filename, 'r+') as f:
+        pin_config = json.load(f)
+        if tag in pin_config:
+            value = pin_config[tag]
+            logging.info(f'{tag}:{value} found in JSON file')
+            pin_config[tag]["in_use"] = True
+            f.seek(0) # should reset file position to the beginning.
+            json.dump(pin_config, f, indent=4)
+            f.truncate() # remove remaining part
+            logging.info(f'{tag} marked as in use')
+
+    return value
+
+
+def reset(pin_config_filename):
+    """Resets the pin configuration JSON file"""
+    logging.info('Resetting config')
+    with open(pin_config_filename, 'w') as f:
+        f.write("{}")
+        logging.info('Config reset')
+
+
+def clear_unused(pin_config_filename):
+    """Identify and remove unused key-value pairs
```

Comment:
so when is the user supposed to be calling this? if it can be called on startup then maybe it would be wise to make a python script that runs on startup and have it call this? because I agree, there should be some way of auto-clearing unused pins.

Reply:
Yeah I just had it in there cause you said it should be there, but there is currently nothing that defines when a pin is "in use"... thoughts?

Reply:
it should go out of use when it's not being used by a currently running program. idc how you do this, but to me this could be done by having a script run on startup and look at what pins are in use then remove those which are not. alternatively, you could create a script that runs when you call the "remove program" from the command line utility. (This might actually be easier)